crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -318,9 +318,10 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.types.has_deferred |= inference.has_deferred;
     }
 
-    /// Are we currently inferring types in a stub file?
-    fn is_stub(&self) -> bool {
-        self.file.is_stub(self.db.upcast())
+    /// Are we currently inferring types in file with deferred types?
+    /// This is true for stub files and files with `__future__.annotations`
+    fn are_all_types_deferred(&self) -> bool {
+        self.file.is_stub(self.db.upcast()) || self.index.has_future_annotations()
```

Comment:
nit: checking `has_future_annotations()` is slightly cheaper, maybe let's do that first?